File: PairWithTheGivenDifference.py
def findPair(arr, size, n):
    mpp = {} # Tạo một từ điển mpp để lưu trữ số lần xuất hiện của mỗi phần tử
    # Duyệt qua mảng để kiểm tra các phần tử trùng lặp khi n == 0
    for i in range(size):
        # Nếu phần tử arr[i] đã có trong mpp
        if arr[i] in mpp:
            mpp[arr[i]] += 1
            # Nếu n == 0 và phần tử xuất hiện nhiều hơn 1 lần, ta tìm thấy cặp
            if n == 0 and mpp[arr[i]] > 1:
                print("Cặp tìm thấy: (" + str(arr[i]) + ", " + str(arr[i]) + ")")
                return True
        else:
            mpp[arr[i]] = 1  # Nếu phần tử chưa có trong mpp, thêm nó vào
    # Nếu không có cặp trùng lặp và n == 0, trả về False
    if n == 0:
        return False
    # Duyệt lại mảng lần nữa để tìm cặp có hiệu là n
    for i in range(size):
        # Kiểm tra nếu phần tử arr[i] + n có trong mpp
        if n + arr[i] in mpp:
            print("Cặp tìm thấy: (" + str(arr[i]) + ", " + str(n + arr[i]) + ")")
            return True
    print("Không tìm thấy cặp nào")  # Nếu không tìm thấy cặp nào, in ra thông báo
    return False
arr = [1, 8, 30, 40, 100]  # Mảng đầu vào
size = len(arr)  # Tính độ dài mảng
n = -60  # Giá trị hiệu cần tìm
findPair(arr, size, n)

File: PairWithTheGivenDifference.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX 1000  // Định nghĩa kích thước tối đa cho bảng băm

// Hàm tìm cặp có hiệu bằng n trong mảng chưa sắp xếp
int findPairWithDifference(int arr[], int size, int n) {
    int mpp[MAX] = {0};  // Mảng đánh dấu các phần tử đã gặp

    for (int i = 0; i < size; i++) {
        // Kiểm tra xem có phần tử nào có hiệu với arr[i] bằng n không
        if (mpp[arr[i] + n] > 0) {  // Nếu arr[i] + n đã có trong bảng băm
            printf("Cặp tìm thấy: (%d, %d)\n", arr[i], arr[i] + n);
            return 1;  // Tìm thấy cặp
        }
        if (arr[i] - n >= 0 && mpp[arr[i] - n] > 0) {  // Nếu arr[i] - n đã có trong bảng băm
            printf("Cặp tìm thấy: (%d, %d)\n", arr[i], arr[i] - n);
            return 1;  // Tìm thấy cặp
        }
        
        // Lưu phần tử vào Hash Table
        mpp[arr[i]] = 1;
    }

    printf("Không tìm thấy cặp nào\n");
    return 0;  // Không tìm thấy cặp
}

int main() {
    int n;

    // Nhập số phần tử trong mảng
    printf("Nhập số lượng phần tử trong mảng: ");
    scanf("%d", &n);

    int arr[n];

    // Nhập các phần tử trong mảng
    printf("Nhập các phần tử trong mảng: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    // Nhập giá trị hiệu cần tìm
    int diff;
    printf("Nhập giá trị hiệu cần tìm (có thể là số âm): ");
    scanf("%d", &diff);

    // Gọi hàm tìm cặp có hiệu bằng diff
    findPairWithDifference(arr, n, diff);

    return 0;
}
